File: src/main/java/com/aws/emr/spark/cdc/SparkCDCMirror.java
package com.aws.emr.spark.cdc;

import com.aws.emr.common.JobConfig;
import java.io.IOException;
import java.util.Map;
import java.util.concurrent.TimeoutException;
import org.apache.spark.sql.*;
import org.apache.spark.sql.streaming.StreamingQueryException;

/**
 * An example of the MERGE INTO CDC "mirror" pattern: a Spark batch pipeline that deduplicates the
 * {@code accounts_changelog} table (produced by {@link SparkLogChange}) and merges the latest change
 * per key into the {@code accounts_mirror} table.
 *
 * <p>The {@code accounts_mirror} table is created as an Iceberg format-version 3 (v3) table, so the
 * deletes produced by the MERGE are written as deletion vectors. The Spark session, catalog and run
 * environment are selected through {@link JobConfig} {@code key=value} arguments; see
 * {@link JobConfig#usage()}.
 *
 * @author acmanjon@amazon.com
 */
public class SparkCDCMirror {

  public static void main(String[] args)
      throws IOException, TimeoutException, StreamingQueryException {

    JobConfig cfg = JobConfig.fromArgs(args);
    SparkSession spark = cfg.buildSession("CDCMirrorMerge");

    spark.sql("CREATE DATABASE IF NOT EXISTS " + JobConfig.DATABASE);
    spark.sql("USE " + JobConfig.DATABASE);
    // Merge-on-read mirror by default (the MERGE's deletes/updates become deletion vectors on v3);
    // override with mode=/fv=/fileformat= like every other example.
    spark.sql(
        cfg.createTableDdl(
            "accounts_mirror",
            CdcSql.MIRROR_COLUMNS_DDL,
            CdcSql.MIRROR_PARTITION_DDL,
            JobConfig.Mode.MOR,
            Map.of()));

    // We only scan changes from the last day so we don't deduplicate over the whole (huge) changelog.
    // Dedup keeps the highest source sequence per key (deterministic) and the MERGE guards updates and
    // deletes with c.seq >= a.seq so a stale change can never overwrite newer state. The 1-day filter
    // is a coarse late-arrival window; for exact incremental reads see SparkIncrementalPipeline and
    // https://tabular.io/apache-iceberg-cookbook/data-engineering-incremental-processing/
    spark.sql(
        CdcSql.mirrorMerge(
            "accounts_mirror",
            "(SELECT * FROM accounts_changelog WHERE last_updated > current_timestamp() - INTERVAL 1 DAY) src"));
  }
}

File: python/src/iceberg_streaming/cdc/spark_cdc_mirror.py
"""CDC mirror MERGE pattern (batch).

PySpark counterpart of ``com.aws.emr.spark.cdc.SparkCDCMirror``. Deduplicates the
``accounts_changelog`` table (highest source ``seq`` per key) and merges it into the
``accounts_mirror`` table with the shared, guarded MERGE (see :mod:`iceberg_streaming.cdc._sql`).
The mirror is merge-on-read by default, so the deletes are written as deletion vectors on v3;
override with ``mode=``/``fv=``/``fileformat=`` like every other example.
"""

from __future__ import annotations

import sys

from iceberg_streaming.cdc import _sql
from iceberg_streaming.common import DATABASE, JobConfig, Mode

# Only scan the last day of changes so we don't deduplicate over the whole changelog. Dedup keeps the
# highest source sequence per key and the MERGE guards updates/deletes with c.seq >= a.seq.
_MERGE = _sql.mirror_merge(
    "accounts_mirror",
    "(SELECT * FROM accounts_changelog WHERE last_updated > current_timestamp() - INTERVAL 1 DAY) src",
)


def main(argv: list[str] | None = None) -> None:
    cfg = JobConfig.from_args(argv if argv is not None else sys.argv[1:])
    spark = cfg.build_session("PySparkCDCMirrorMerge")

    spark.sql(f"CREATE DATABASE IF NOT EXISTS {DATABASE}")
    spark.sql(f"USE {DATABASE}")
    spark.sql(
        cfg.create_table_ddl(
            "accounts_mirror", _sql.MIRROR_COLUMNS_DDL, _sql.MIRROR_PARTITION_DDL, Mode.MOR
        )
    )
    spark.sql(_MERGE)


if __name__ == "__main__":
    main()
